# ASSIGNMENT NO 10  Association Rule

In [19]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt


In [20]:
df = pd.read_csv("Online retail.csv")                                                                                                                                                                                                                                                   


In [21]:
df


Products
0     shrimp,almonds,avocado,vegetables mix,green gr...
1                                burgers,meatballs,eggs
2                                               chutney
3                                        turkey,avocado
4     mineral water,milk,energy bar,whole wheat rice...
...                                                 ...
7496                      butter,light mayo,fresh bread
7497  burgers,frozen vegetables,eggs,french fries,ma...
7498                                            chicken
7499                                 escalope,green tea
7500    eggs,frozen smoothie,yogurt cake,low fat yogurt

[7501 rows x 1 columns]

**Data Preprocessing:**
----------------------------------------------------------------------------------
Pre-process the dataset to ensure it is suitable for Association rules, this may include handling missing values, removing duplicates, and converting the data to appropriate format.  


In [22]:
df.drop_duplicates(inplace=True)


In [23]:
def txt_split(txt):
    return txt.split(',')


In [24]:
data = df['Products'].apply(txt_split)


In [25]:
data


0       [shrimp, almonds, avocado, vegetables mix, gre...
1                              [burgers, meatballs, eggs]
2                                               [chutney]
3                                       [turkey, avocado]
4       [mineral water, milk, energy bar, whole wheat ...
                              ...                        
7493    [burgers, salmon, pancakes, french fries, froz...
7494    [turkey, burgers, dessert wine, shrimp, pasta,...
7496                    [butter, light mayo, fresh bread]
7497    [burgers, frozen vegetables, eggs, french frie...
7500    [eggs, frozen smoothie, yogurt cake, low fat y...
Name: Products, Length: 5176, dtype: object

In [26]:
data.isna().sum()
# No missing Value


0

In [27]:
from mlxtend.preprocessing import TransactionEncoder


In [28]:
te = TransactionEncoder()
encoded_df = te.fit_transform(data)
encoded_df


array([[False,  True,  True, ...,  True, False, False],
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False],
       ...,
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False,  True, False]])

In [29]:
te.columns_


[' asparagus',
 'almonds',
 'antioxydant juice',
 'asparagus',
 'avocado',
 'babies food',
 'bacon',
 'barbecue sauce',
 'black tea',
 'blueberries',
 'body spray',
 'bramble',
 'brownies',
 'bug spray',
 'burger sauce',
 'burgers',
 'butter',
 'cake',
 'candy bars',
 'carrots',
 'cauliflower',
 'cereals',
 'champagne',
 'chicken',
 'chili',
 'chocolate',
 'chocolate bread',
 'chutney',
 'cider',
 'clothes accessories',
 'cookies',
 'cooking oil',
 'corn',
 'cottage cheese',
 'cream',
 'dessert wine',
 'eggplant',
 'eggs',
 'energy bar',
 'energy drink',
 'escalope',
 'extra dark chocolate',
 'flax seed',
 'french fries',
 'french wine',
 'fresh bread',
 'fresh tuna',
 'fromage blanc',
 'frozen smoothie',
 'frozen vegetables',
 'gluten free bar',
 'grated cheese',
 'green beans',
 'green grapes',
 'green tea',
 'ground beef',
 'gums',
 'ham',
 'hand protein bar',
 'herb & pepper',
 'honey',
 'hot dogs',
 'ketchup',
 'light cream',
 'light mayo',
 'low fat yogurt',
 'magazines',
 'mashe

In [30]:
data = pd.DataFrame(encoded_df, columns=te.columns_)
data


asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0          False     True               True      False     True        False   
1          False    False              False      False    False        False   
2          False    False              False      False    False        False   
3          False    False              False      False     True        False   
4          False    False              False      False    False        False   
...          ...      ...                ...        ...      ...          ...   
5171       False    False              False      False    False        False   
5172       False    False              False      False    False        False   
5173       False    False              False      False    False        False   
5174       False    False              False      False    False        False   
5175       False    False              False      False    False        False   

      bacon  barbecue sauce  black tea  blueberries  ...  turkey  \
0     False           False      False        False  ...   False   
1     False           False      False        False  ...   False   
2     False           False      False        False  ...   False   
3     False           False      False        False  ...    True   
4     False           False      False        False  ...   False   
...     ...             ...        ...          ...  ...     ...   
5171  False           False      False        False  ...   False   
5172  False           False      False        False  ...    True   
5173  False           False      False        False  ...   False   
5174  False           False      False        False  ...   False   
5175  False           False      False        False  ...   False   

      vegetables mix  water spray  white wine  whole weat flour  \
0               True        False       False              True   
1              False        False       False             False   
2              False        False       False             False   
3              False        False       False             False   
4              False        False       False             False   
...              ...          ...         ...               ...   
5171           False        False       False             False   
5172           False        False       False             False   
5173           False        False       False             False   
5174           False        False       False             False   
5175           False        False       False             False   

      whole wheat pasta  whole wheat rice   yams  yogurt cake  zucchini  
0                 False             False   True        False     False  
1                 False             False  False        False     False  
2                 False             False  False        False     False  
3                 False             False  False        False     False  
4                 False              True  False        False     False  
...                 ...               ...    ...          ...       ...  
5171              False             False  False        False     False  
5172              False              True  False        False     False  
5173              False             False  False        False     False  
5174              False             False  False        False     False  
5175              False             False  False         True     False  

[5176 rows x 120 columns]

In [31]:
data.replace([True,False],[1,0], inplace=True)


C:\Users\Yogesh\AppData\Local\Temp\ipykernel_7788\2274451266.py:1: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  data.replace([True,False],[1,0], inplace=True)


In [32]:
data.shape


(5176, 120)

**Association Rule Mining:**
- Implement an Apriori algorithm using tool like python with libraries such as Pandas and Mlxtend etc.
-	 Apply association rule mining techniques to the pre-processed dataset to discover interesting relationships between products purchased together.
-	Set appropriate threshold for support, confidence and lift to extract meaning full rules.


In [33]:
from mlxtend.frequent_patterns import association_rules,apriori


In [34]:
scores = apriori(data,min_support=0.05, use_colnames=True)
scores


C:\Users\Yogesh\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\mlxtend\frequent_patterns\fpcommon.py:109: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support                            itemsets
0   0.113794                           (burgers)
1   0.103555                              (cake)
2   0.054869                         (champagne)
3   0.083849                           (chicken)
4   0.205178                         (chocolate)
5   0.060665                           (cookies)
6   0.071870                       (cooking oil)
7   0.208076                              (eggs)
8   0.083849                          (escalope)
9   0.192620                      (french fries)
10  0.059312                       (fresh bread)
11  0.084815                   (frozen smoothie)
12  0.129830                 (frozen vegetables)
13  0.070131                     (grated cheese)
14  0.169243                         (green tea)
15  0.135819                       (ground beef)
16  0.066461                     (herb & pepper)
17  0.065301                             (honey)
18  0.095054                    (low fat yogurt)
19  0.170015                              (milk)
20  0.299845                     (mineral water)
21  0.087713                         (olive oil)
22  0.125193                          (pancakes)
23  0.057573                            (salmon)
24  0.098918                            (shrimp)
25  0.070904                              (soup)
26  0.229521                         (spaghetti)
27  0.091963                          (tomatoes)
28  0.083269                            (turkey)
29  0.077859                  (whole wheat rice)
30  0.073223          (mineral water, chocolate)
31  0.055835              (spaghetti, chocolate)
32  0.070131               (mineral water, eggs)
33  0.051391                   (spaghetti, eggs)
34  0.050425  (mineral water, frozen vegetables)
35  0.058733        (mineral water, ground beef)
36  0.055835            (ground beef, spaghetti)
37  0.067813               (milk, mineral water)
38  0.050039                   (milk, spaghetti)
39  0.085008          (mineral water, spaghetti)

In [35]:
rules = association_rules(scores, metric="lift", min_threshold=1)
rules[['antecedents', 'consequents', 'support', 'confidence', 'lift']]


antecedents          consequents   support  confidence      lift
0       (mineral water)          (chocolate)  0.073223    0.244201  1.190193
1           (chocolate)      (mineral water)  0.073223    0.356874  1.190193
2           (spaghetti)          (chocolate)  0.055835    0.243266  1.185635
3           (chocolate)          (spaghetti)  0.055835    0.272128  1.185635
4       (mineral water)               (eggs)  0.070131    0.233892  1.124070
5                (eggs)      (mineral water)  0.070131    0.337047  1.124070
6           (spaghetti)               (eggs)  0.051391    0.223906  1.076078
7                (eggs)          (spaghetti)  0.051391    0.246982  1.076078
8       (mineral water)  (frozen vegetables)  0.050425    0.168170  1.295310
9   (frozen vegetables)      (mineral water)  0.050425    0.388393  1.295310
10      (mineral water)        (ground beef)  0.058733    0.195876  1.442184
11        (ground beef)      (mineral water)  0.058733    0.432432  1.442184
12        (ground beef)          (spaghetti)  0.055835    0.411095  1.791102
13          (spaghetti)        (ground beef)  0.055835    0.243266  1.791102
14               (milk)      (mineral water)  0.067813    0.398864  1.330231
15      (mineral water)               (milk)  0.067813    0.226160  1.330231
16               (milk)          (spaghetti)  0.050039    0.294318  1.282316
17          (spaghetti)               (milk)  0.050039    0.218013  1.282316
18      (mineral water)          (spaghetti)  0.085008    0.283505  1.235204
19          (spaghetti)      (mineral water)  0.085008    0.370370  1.235204

In [36]:
rules.sort_values(by = 'lift', ascending=False)


antecedents          consequents  antecedent support  \
13          (spaghetti)        (ground beef)            0.229521   
12        (ground beef)          (spaghetti)            0.135819   
11        (ground beef)      (mineral water)            0.135819   
10      (mineral water)        (ground beef)            0.299845   
15      (mineral water)               (milk)            0.299845   
14               (milk)      (mineral water)            0.170015   
8       (mineral water)  (frozen vegetables)            0.299845   
9   (frozen vegetables)      (mineral water)            0.129830   
17          (spaghetti)               (milk)            0.229521   
16               (milk)          (spaghetti)            0.170015   
18      (mineral water)          (spaghetti)            0.299845   
19          (spaghetti)      (mineral water)            0.229521   
1           (chocolate)      (mineral water)            0.205178   
0       (mineral water)          (chocolate)            0.299845   
2           (spaghetti)          (chocolate)            0.229521   
3           (chocolate)          (spaghetti)            0.205178   
5                (eggs)      (mineral water)            0.208076   
4       (mineral water)               (eggs)            0.299845   
7                (eggs)          (spaghetti)            0.208076   
6           (spaghetti)               (eggs)            0.229521   

    consequent support   support  confidence      lift  leverage  conviction  \
13            0.135819  0.055835    0.243266  1.791102  0.024661    1.141988   
12            0.229521  0.055835    0.411095  1.791102  0.024661    1.308326   
11            0.299845  0.058733    0.432432  1.442184  0.018008    1.233606   
10            0.135819  0.058733    0.195876  1.442184  0.018008    1.074686   
15            0.170015  0.067813    0.226160  1.330231  0.016835    1.072553   
14            0.299845  0.067813    0.398864  1.330231  0.016835    1.164718   
8             0.129830  0.050425    0.168170  1.295310  0.011496    1.046091   
9             0.299845  0.050425    0.388393  1.295310  0.011496    1.144778   
17            0.170015  0.050039    0.218013  1.282316  0.011017    1.061380   
16            0.229521  0.050039    0.294318  1.282316  0.011017    1.091822   
18            0.229521  0.085008    0.283505  1.235204  0.016187    1.075345   
19            0.299845  0.085008    0.370370  1.235204  0.016187    1.112010   
1             0.299845  0.073223    0.356874  1.190193  0.011701    1.088674   
0             0.205178  0.073223    0.244201  1.190193  0.011701    1.051632   
2             0.205178  0.055835    0.243266  1.185635  0.008742    1.050332   
3             0.229521  0.055835    0.272128  1.185635  0.008742    1.058537   
5             0.299845  0.070131    0.337047  1.124070  0.007741    1.056115   
4             0.208076  0.070131    0.233892  1.124070  0.007741    1.033698   
7             0.229521  0.051391    0.246982  1.076078  0.003633    1.023189   
6             0.208076  0.051391    0.223906  1.076078  0.003633    1.020397   

    zhangs_metric  
13       0.573260  
12       0.511102  
11       0.354795  
10       0.437914  
15       0.354566  
14       0.299103  
8        0.325620  
9        0.262000  
17       0.285745  
16       0.265259  
18       0.271965  
19       0.247141  
1        0.201051  
0        0.228235  
2        0.203212  
3        0.196988  
5        0.139377  
4        0.157645  
7        0.089275  
6        0.091760

**Analysis and Interpretation:**
- Analyse the generated rules to identify interesting patterns and relationships between the products.
- Interpret the results and provide insights into customer purchasing behavior based on the discovered rules.

--- 
### Rules Summary:

1. **(spaghetti → ground beef)**
   - **Support**: 0.0558 (5.58% of transactions include spaghetti and ground beef together)
   - **Confidence**: 0.243 (24.3% of transactions with spaghetti also have ground beef)
   - **Lift**: 1.79 (Spaghetti is 79% more likely to be bought with ground beef than randomly)
   - **Interpretation**: There’s a moderate association between spaghetti and ground beef, which could be useful for cross-promotion strategies.

2. **(ground beef → spaghetti)**
   - **Support**: 0.0558
   - **Confidence**: 0.411 (41.1% of ground beef buyers also buy spaghetti)
   - **Lift**: 1.79
   - **Interpretation**: Ground beef buyers are more likely to buy spaghetti, which could suggest a strong relation (like spaghetti bolognese).

3. **(ground beef → mineral water)**
   - **Support**: 0.0587
   - **Confidence**: 0.432 (43.2% of ground beef buyers also buy mineral water)
   - **Lift**: 1.44
   - **Interpretation**: There is a fairly strong connection between ground beef and mineral water. Mineral water could be used as an upsell or complementary product for beef buyers.

4. **(mineral water → ground beef)**
   - **Support**: 0.0587
   - **Confidence**: 0.195 (19.6% of mineral water buyers buy ground beef)
   - **Lift**: 1.44
   - **Interpretation**: This rule has a lower confidence, suggesting fewer mineral water buyers choose ground beef, but the lift suggests it’s still a meaningful relationship.

5. **(mineral water → milk)**
   - **Support**: 0.0678 (6.78% of transactions include both mineral water and milk)
   - **Confidence**: 0.226 (22.6% of mineral water buyers also buy milk)
   - **Lift**: 1.33 (There’s a 33% increased chance that mineral water buyers also purchase milk)
   - **Interpretation**: This shows a moderate relationship between mineral water and milk. They could be placed together in promotions.

6. **(milk → mineral water)**
   - **Support**: 0.0678
   - **Confidence**: 0.399 (39.9% of milk buyers also buy mineral water)
   - **Lift**: 1.33
   - **Interpretation**: Milk buyers have a relatively high likelihood of buying mineral water, which could indicate complementary consumption patterns.

7. **(mineral water → frozen vegetables)**
   - **Support**: 0.0504
   - **Confidence**: 0.168 (16.8% of mineral water buyers purchase frozen vegetables)
   - **Lift**: 1.29
   - **Interpretation**: This is a weaker relationship, but the lift still indicates that mineral water buyers are more likely to purchase frozen vegetables than random customers.

8. **(frozen vegetables → mineral water)**
   - **Support**: 0.0504
   - **Confidence**: 0.388 (38.8% of frozen vegetable buyers purchase mineral water)
   - **Lift**: 1.29
   - **Interpretation**: Frozen vegetable buyers have a higher likelihood of purchasing mineral water, which could be used to target health-conscious shoppers.

9. **(spaghetti → milk)**
   - **Support**: 0.0500
   - **Confidence**: 0.218 (21.8% of spaghetti buyers also buy milk)
   - **Lift**: 1.28
   - **Interpretation**: Spaghetti and milk show a moderate relationship, with some potential for promotions or bundling, but it’s not as strong as other relationships in the data.

---

### Conclusion:
- The most interesting associations seem to be between **ground beef and spaghetti** (with high confidence and lift values), suggesting a strong relationship between these two items.
- **Mineral water** appears frequently with other products like ground beef, milk, and frozen vegetables. This indicates that mineral water is often bought in combination with various other items, possibly making it a good candidate for **bundling** or **cross-promotions**.
- **Frozen vegetables** and **mineral water** are moderately associated, suggesting an opportunity to target health-conscious buyers with joint offers.

These insights can be used for improving **product placements**, **advertising strategies**, or designing **special promotions**.